kommandir/bin/adept_openstack.py

```diff
@@ -0,0 +1,720 @@
+        networks = self.response_json['addresses'].keys()
+        self.raise_if(len(networks) > 1,
+                      ValueError,
+                      "Found %s to be on more than one network" % name)
+        self.raise_if(len(networks) == 0,
+                      ValueError,
+                      "Found %s to be on no networks" % name)
+        try:
+            ifaces = self.response_json['addresses'][networks[0]]
+            floating_iface = self.child_search('OS-EXT-IPS:type', 'floating',
+                                               alt_list=ifaces)
+            return floating_iface['addr']
+        except (KeyError, IndexError), xcept:
+            self.raise_if(True,
+                          xcept,
+                          "Found %s to be on a network"
+                          " without a floating IP address" % name)
+
+    def server_delete(self, name=None, uuid=None):
+        """
+        Cache list of servers, try to delete server by name or uuid
+
+        :param name: Optional, name of server to retrieve IP from
+        :param uuid: Optional, ID of server to retrieve IP from
+        :returns: True if server delete request successful
+        """
+        try:
+            server_details = self.server(name=name, uuid=uuid)
+        except IndexError:
+            return True # Server doesn't exist
+        try:
+            self.compute_request('/servers/%s' % server_details['id'],
+                                 method='delete')
+            return True
+        # This can fail for any number of reasons, can't list them all
+        # pylint: disable=W0702
+        except:
+            return False
+
+    def floating_ip(self):
+        """
+        Cache list of floating IPs, return first un-assigned or None
+
+        :returns: IP address string or None
+        """
+        self.service_request('network', '/v2.0/floatingips', 'floatingips')
+        try:
+            return self.child_search('status', value='DOWN')["floating_ip_address"]
+        except (KeyError, IndexError):
+            return None
+
+
+class TimeoutAction(object):
+    """ABC callable that raises an exception on timeout"""
+
+    timeout = 300.0  # (seconds)
+    time_out_at = None  # absolute
+    timeout_exception = RuntimeError
+
+    def __init__(self, *args, **dargs):
+        del args
+        del dargs
+        _ = self
+
+    def __str__(self, *args, **dargs):
+        return ("%s(*%s, **%s) after %0.2f"
+                % (self.__class__.__name__, args, dargs, self.timeout))
+
+    def __call__(self, *args, **dargs):
+        result = None
+        start = time.time()
+        if self.time_out_at is None:
+            self.time_out_at = start + self.timeout
+        while True:
+            if time.time() >= self.time_out_at:
+                raise self.timeout_exception(str(self))
+            result = self.action(*args, **dargs)
+            if self.done():
+                break
+        return result
```

Comment:
...Seems like I could just nuke ``action()`` and move all that code into ``__init__()``.  There are a few subclasses that would need minor re-jiggering, but it could work.  Giving it a try...